cpp/inc/bond/core/capped_allocator.h

```diff
@@ -0,0 +1,680 @@
+            friend class value_or_reference;
+
+            boost::optional<T> _value;
+            std::reference_wrapper<T> _ref;
+        };
+
+
+        template <typename T>
+        class value_or_reference<T&>
+        {
+        public:
+            value_or_reference(T& value)
+                : _ref{ value }
+            {}
+
+            value_or_reference(std::reference_wrapper<T> ref)
+                : value_or_reference{ ref.get() }
+            {}
+
+            value_or_reference(boost::reference_wrapper<T> ref)
+                : value_or_reference{ ref.get() }
+            {}
+
+            T& get() const
+            {
+                return _ref;
+            }
+
+        private:
+            std::reference_wrapper<T> _ref;
+        };
+
+
+        /// @brief Helper type that holds an allocator.
+        ///
+        /// @remarks Applies empty-base-optimization when possible.
+        template <typename Alloc, typename Enable = void>
+        class allocator_holder : private Alloc
+        {
+        public:
+            allocator_holder() = default;
+
+            allocator_holder(const Alloc& alloc)
+                : Alloc{ alloc }
+            {}
+
+            const Alloc& get() const
+            {
+                return *this;
+            }
+
+            Alloc& get()
+            {
+                return *this;
+            }
+        };
+
+#if defined(_MSC_VER) && _MSC_VER < 1900
+#pragma warning(push)
+#pragma warning(disable: 4510)
+#endif
+        template <typename Alloc>
+        class allocator_holder<Alloc, typename boost::disable_if<std::is_empty<Alloc>>::type>
+        {
+        public:
+            allocator_holder() = default;
+
+            allocator_holder(const Alloc& alloc)
+                : _alloc{ alloc }
+            {}
+
+            const Alloc& get() const
+            {
+                return _alloc;
+            }
+
+            Alloc& get()
+            {
+                return _alloc;
+            }
+
+        private:
+            Alloc _alloc;
+        };
+#if defined(_MSC_VER) && _MSC_VER < 1900
+#pragma warning(pop)
+#endif
+
+
+        /// @brief Helper base class for counters.
+        template <typename T>
+        class capped_allocator_counter_base
+        {
+        public:
+            BOOST_STATIC_ASSERT(std::is_integral<T>::value);
+
+            using value_type = T;
+
+            explicit capped_allocator_counter_base(T max_value) BOND_NOEXCEPT
+                : _max_value{ max_value }
+            {}
+
+            capped_allocator_counter_base(const capped_allocator_counter_base& other) = delete;
+
+            T max_value() const BOND_NOEXCEPT
+            {
+                return _max_value;
+            }
+
+        private:
+            const T _max_value;
+        };
+
+    } // namespace detail
+
+
+    template <typename T>
+    class capped_allocator_counter<T, /*SingleThreaded*/ true> : public detail::capped_allocator_counter_base<T>
```

Comment:
I feel like the SingleThreaded boolean should be replaced with a counter_policy concept that can have single_threaded and multi_threaded implementations.

Reply:
How about having `capped_allocator_single_threaded_counter<T>`, `capped_allocator_multi_threaded_counter<T>` and `capped_allocator_shared_counter<Counter>` which takes one of the two?

Reply:
Possibly. I'd love to eliminate or reduce the presence of boolean SingleThreaded and simplify the implementation. of capped_allocator_counter.